guidance/models/transformers/_transformers_phi3v.py

```diff
@@ -0,0 +1,218 @@
+import logging
+import io
+import os
+from typing import Optional
+
+import numpy as np
+import torch
+import llguidance
+from transformers import AutoModelForCausalLM, AutoProcessor
+
+from guidance._parser import TokenParser, process_grammar, process_prompt
+from guidance.models._model import (
+    Engine,
+    Model,
+    modality_pattern,
+    Modality
+)
+from guidance.models.transformers._transformers import TransformersTokenizer
+
+try:
+    from PIL import Image
+    has_pillow = True
+except ModuleNotFoundError:
+    has_pillow = False
+
+logger = logging.getLogger(__name__)
+
+
+class TransformersPhi3VisionEngine(Engine):
+    def __init__(
+        self,
+        model="microsoft/Phi-3-vision-128k-instruct",
+        compute_log_probs=False,
+        **kwargs,
+    ):
+        if not has_pillow:
+            raise Exception("Please install pillow with `pip install pillow` to use Phi 3 Vision")
+        self.model_name = model
+        # Initialize the underlying Phi 3 Vision model
+        self.model_obj = AutoModelForCausalLM.from_pretrained(model, **kwargs)
+        self.device = self.model_obj.device
+
+        # Processor handles tokenization and image processing
+        self.processor = AutoProcessor.from_pretrained(self.model_name, trust_remote_code=True)
+        self.tokenizer = TransformersTokenizer(model, self.processor.tokenizer, sp_whitespace=True)
+        super().__init__(self.tokenizer, compute_log_probs)
+
+        # Cache for past key values
+        self._past_key_values = None
+        self._cached_token_ids: list[int] = []
+
+
+    def start(self, prompt, grammar, media: dict, ensure_bos_token=True) -> TokenParser:
+        if isinstance(prompt, bytes):
+            prompt = prompt.decode("utf-8")
+        elif isinstance(prompt, str):
+            prompt = prompt
```

Comment:
Maybe missing a str -> byte conversion, or skip conditional altogether?

Reply:
Agreed. Side note, but I want us to be a bit more careful about strings vs bytes in general -- maybe some annotations would help (as well as appropriately typing the base class to either be one or the other or a `Union[str,bytes]` if that makes the Liskov problem easier...)